compiler/codegen/codegen.cpp

```diff
@@ -1117,58 +1117,150 @@ static void codegen_header_compilation_config() {
   if (cfgfile.fptr != NULL) {
     FILE* save_cfile = gGenInfo->cfile;
 
-    gGenInfo->cfile = cfgfile.fptr;
+    fprintf(cfgfile.fptr, "/*** %s ***/\n", "Compilation Info");
+    fprintf(cfgfile.fptr, "\n#include <stdio.h>");
+    fprintf(cfgfile.fptr, "\n#include \"chpltypes.h\"\n\n");
 
-    genComment("Compilation Info");
+    if (llvmCodegen) {
+#ifdef HAVE_LLVM
+      genGlobalString("chpl_compileCommand", compileCommand);
+      genGlobalString("chpl_compileVersion", compileVersion);
+      genGlobalString("chpl_compileDirectory", getCwd());
+      if (strcmp(saveCDir, "") != 0) {
+        char *actualPath = realpath(saveCDir, NULL);
+        genGlobalString("chpl_saveCDir", actualPath);
+      } else {
+        genGlobalString("chpl_saveCDir", "");
+      }
+
+      genGlobalString("CHPL_HOME", CHPL_HOME);
+
+      genGlobalInt("CHPL_STACK_CHECKS", !fNoStackChecks, false);
+      genGlobalInt("CHPL_CACHE_REMOTE", fCacheRemote, false);
+
+      for (std::map<std::string, const char*>::iterator env=envMap.begin(); env!=envMap.end(); ++env) {
+        if (env->first != "CHPL_HOME") {
+          genGlobalString(env->first.c_str(), env->second);
+        }
+      }
+
+      llvm::FunctionType* programAboutType;
+      llvm::Function* programAboutFunc;
+      if ((programAboutFunc = getFunctionLLVM("chpl_program_about"))) {
+        programAboutType = programAboutFunc->getFunctionType();
+      } else {
+        programAboutType = llvm::FunctionType::get(
+          llvm::Type::getVoidTy(gGenInfo->module->getContext()), false
+        );
+        programAboutFunc = llvm::Function::Create(
+          programAboutType, llvm::Function::ExternalLinkage, "chpl_program_about", gGenInfo->module
+        );
+      }
+
+      llvm::BasicBlock* programAboutBlock = llvm::BasicBlock::Create(
+        gGenInfo->module->getContext(), "entry", programAboutFunc
+      );
+      gGenInfo->irBuilder->SetInsertPoint(programAboutBlock);
+
+      llvm::FunctionType* printfType = llvm::FunctionType::get(
+        llvm::Type::getInt32Ty(gGenInfo->module->getContext()), {}, true
+      );
+      llvm::Function* printfFunc = llvm::cast<llvm::Function>(
+        gGenInfo->module->getOrInsertFunction("printf", printfType)
+      );
+
+      std::vector<llvm::Value *> compilationCommandArgs(2);
+      compilationCommandArgs[0] = gGenInfo->irBuilder->CreateLoad(
+        new_CStringSymbol("Compilation command: %s\n")->codegen().val);
+      compilationCommandArgs[1] = gGenInfo->irBuilder->CreateLoad(
+        new_CStringSymbol(compileCommand)->codegen().val);
+      gGenInfo->irBuilder->CreateCall(printfFunc, compilationCommandArgs);
```

Comment:
So the global variables should work for both C and LLVM with genGlobalString etc.
I think LLVM / C needs some differences here to code generate within the particular function.
However, I think it might be cleaner if you added a codegenCallPrintf(GenRet a, ...) (maybe versions accepting 2 or 3 args) and used codegenValue() instead of CreateLoad. You cuold add codegenCallPrintf to compiler/codegen/expr.cpp and have it call codegenCall; or you could just handle it directly in this file. Either way though I think it's worth factoring it out.